### PR TITLE
[Yaft][0.0.4] - New Package

### PR DESCRIPTION
**Pull request by timower**
Adds a new terminal emulator based on Yaft.

### CODE REVIEW

**Comment by LinusCDE** [comment]
Hi. First thanks for the contribution. Having a faster on-device terminal emulator sounds really cool!

I failed to build the package to test it though. It's the first time of me building something with the new python based build-system. Though the error seems to be because of a cmake configuration problem.

<details>
<summary>Output of running <code>FLAGS=--verbose make yaft</code></summary>

```shell
╭─linus@HomeLX /tmp/tol1/toltec  ‹add_yaft›
╰─$ FLAGS=--verbose make yaft
./scripts/package_build.py --verbose "yaft"
[   DEBUG] docker.utils.config: Trying paths: ['/home/linus/.docker/config.json', '/home/linus/.dockercfg']
[   DEBUG] docker.utils.config: No config file found
[   DEBUG] docker.utils.config: Trying paths: ['/home/linus/.docker/config.json', '/home/linus/.dockercfg']
[   DEBUG] docker.utils.config: No config file found
[   DEBUG] urllib3.connectionpool: http://localhost:None "GET /version HTTP/1.1" 200 833
[    INFO] toltec.builder: yaft: Fetching source files
[   DEBUG] urllib3.connectionpool: Starting new HTTPS connection (1): github.com:443
[   DEBUG] urllib3.connectionpool: https://github.com:443 "GET /timower/rM2-stuff/archive/0a5ff42551dab208c1937d9b10396f30e8e982ba.zip HTTP/1.1" 302 156
[   DEBUG] urllib3.connectionpool: Starting new HTTPS connection (1): codeload.github.com:443
[   DEBUG] urllib3.connectionpool: https://codeload.github.com:443 "GET /timower/rM2-stuff/zip/0a5ff42551dab208c1937d9b10396f30e8e982ba HTTP/1.1" 200 None
[    INFO] toltec.builder: yaft: Skipping prepare (nothing to do)
[    INFO] toltec.builder: yaft: Building artifacts
[   DEBUG] urllib3.connectionpool: http://localhost:None "POST /v1.41/containers/create HTTP/1.1" 201 88
[   DEBUG] urllib3.connectionpool: http://localhost:None "GET /v1.41/containers/39723ed22ed9f169b5039976167ce8c4b5454ee12af7bba41e9c474b6aba3d08/json HTTP/1.1" 200 None
[   DEBUG] urllib3.connectionpool: http://localhost:None "POST /v1.41/containers/39723ed22ed9f169b5039976167ce8c4b5454ee12af7bba41e9c474b6aba3d08/start HTTP/1.1" 204 0
[   DEBUG] urllib3.connectionpool: http://localhost:None "GET /v1.41/containers/39723ed22ed9f169b5039976167ce8c4b5454ee12af7bba41e9c474b6aba3d08/logs?stderr=1&stdout=1&timestamps=0&follow=1&tail=all HTTP/1.1" 200 None
[   DEBUG] urllib3.connectionpool: http://localhost:None "GET /v1.41/containers/39723ed22ed9f169b5039976167ce8c4b5454ee12af7bba41e9c474b6aba3d08/json HTTP/1.1" 200 None
[   DEBUG] toltec.builder: yaft: -- The CXX compiler identification is GNU 8.3.0
[   DEBUG] toltec.builder: yaft: -- Check for working CXX compiler: /usr/bin/c++
[   DEBUG] toltec.builder: yaft: -- Check for working CXX compiler: /usr/bin/c++ -- works
[   DEBUG] toltec.builder: yaft: -- Detecting CXX compiler ABI info
[   DEBUG] toltec.builder: yaft: -- Detecting CXX compiler ABI info - done
[   DEBUG] toltec.builder: yaft: -- Detecting CXX compile features
[   DEBUG] toltec.builder: yaft: -- Detecting CXX compile features - done
[   DEBUG] toltec.builder: yaft: -- The C compiler identification is GNU 8.3.0
[   DEBUG] toltec.builder: yaft: -- Check for working C compiler: /usr/bin/cc
[   DEBUG] toltec.builder: yaft: -- Check for working C compiler: /usr/bin/cc -- works
[   DEBUG] toltec.builder: yaft: -- Detecting C compiler ABI info
[   DEBUG] toltec.builder: yaft: -- Detecting C compiler ABI info - done
[   DEBUG] toltec.builder: yaft: -- Detecting C compile features
[   DEBUG] toltec.builder: yaft: -- Detecting C compile features - done
[   DEBUG] toltec.builder: yaft: CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
[   DEBUG] toltec.builder: yaft: Please set them or make sure they are set and tested correctly in the CMake files:
[   DEBUG] toltec.builder: yaft: /src/libs/rMlib/LIBEVDEV_INCLUDE_DIR
[   DEBUG] toltec.builder: yaft: used as include directory in directory /src/libs/rMlib
[   DEBUG] toltec.builder: yaft: LIBEVDEV_LIBRARY
[   DEBUG] toltec.builder: yaft: linked by target "rMlib" in directory /src/libs/rMlib
[   DEBUG] toltec.builder: yaft:
[   DEBUG] toltec.builder: yaft: -- Configuring incomplete, errors occurred!
[   DEBUG] toltec.builder: yaft: See also "/src/build/CMakeFiles/CMakeOutput.log".
[   DEBUG] urllib3.connectionpool: http://localhost:None "POST /v1.41/containers/39723ed22ed9f169b5039976167ce8c4b5454ee12af7bba41e9c474b6aba3d08/wait HTTP/1.1" 200 None
[   DEBUG] urllib3.connectionpool: http://localhost:None "DELETE /v1.41/containers/39723ed22ed9f169b5039976167ce8c4b5454ee12af7bba41e9c474b6aba3d08?v=False&link=False&force=False HTTP/1.1" 204 0
Traceback (most recent call last):
  File "/tmp/tol1/toltec/./scripts/package_build.py", line 33, in <module>
    if not builder.make(
  File "/tmp/tol1/toltec/scripts/toltec/builder.py", line 126, in make
    self._build(adapter, recipe, src_dir)
  File "/tmp/tol1/toltec/scripts/toltec/builder.py", line 264, in _build
    for line in logs:
  File "/tmp/tol1/toltec/scripts/toltec/bash.py", line 412, in run_script_in_container
    raise ScriptError(f"Script exited with code {result['StatusCode']}")
toltec.bash.ScriptError: Script exited with code 1
make: *** [Makefile:62: yaft] Fehler 1
```

</details>

I also tried building the package from source. Seems that I'm missing some nix specific toolchain?

---

@matteodelabre Any reason for the new system to [fail on retris in ci](https://github.com/toltec-dev/toltec/runs/1764571036)? Seems like an incomplete url.

EDIT: Fixed broken formatting of this comment

**Comment by timower** [comment]
Thanks for having a look, indeed `libevdev` seems to be missing from the base toolchain. It is included on the reMarkable though, so this seems like a toolchain error.

File: package/yaft/package
Comment (matteodelabre) [inline package/yaft/package]: This needs to equal the upstream release time (in this case, the commit date).
```suggestion
timestamp=2021-01-25T19:05:40Z
```

File: package/yaft/package
Comment (matteodelabre) [inline package/yaft/package]: I see that the upstream of your fork has a versioning scheme (currently at 0.2.8). Do you plan on following that same versioning scheme? Or having your own numbering?

File: package/yaft/package
Comment (matteodelabre) [inline package/yaft/package:11]: If I’m not mistaken this should be MIT.
```suggestion
license=MIT
```

File: package/yaft/package
Comment (matteodelabre) [inline package/yaft/package]: If you become the maintainer of this new package, you’ll be responsible for keeping it up to date and test new releases. (See https://github.com/toltec-dev/toltec/blob/stable/docs/branches.md.) If you don’t want to take on this duty, one of the repository maintainers can take ownership of the package. Otherwise, please add a contact email in this field.

**Review by matteodelabre — DISMISSED** [review]
Some metadata nits. Thanks for the PR!

File: package/yaft/package
Comment (timower) [reply package/yaft/package]: I won't keep in sync with upstream, my fork diverges too much. Maybe I'll start at 0.3.0, as upstream hasn't had an update in quite some time.

File: package/yaft/package
Comment (timower) [reply package/yaft/package:11]: I re-licensed the code at GPLv3 as my whole repo uses GPL.

Comment (Eeems) [reply package/yaft/package:11]: That's not what https://github.com/timower/rM2-stuff/blob/master/apps/yaft/LICENSE says

File: package/yaft/package
Comment (timower) [reply package/yaft/package]: Ok, If I'm not a maintainer should I set it to `None` for now?

File: package/yaft/package
Comment (Eeems) [reply package/yaft/package:11]: Also, the header of every file has this included comment:

> /* See LICENSE for licence details. */

Comment (timower) [reply package/yaft/package:11]: Whoops, I forgot to remove that file. It should be replaced by https://github.com/timower/rM2-stuff/blob/master/LICENSE

File: package/yaft/package
Comment (Eeems) [reply package/yaft/package]: The question is, do you want to maintain this package or to leave it to one of us? If you don't want to, you'd set it to `None <none@example.org>`

File: package/timower-stuff/package
Comment (Eeems) [inline package/timower-stuff/package]: This should be split into two separate packages that can be updated independently.

**Review by Eeems — DISMISSED** [review]

File: package/timower-stuff/package
Comment (timower) [reply package/timower-stuff/package]: I'm fine either way, but they'll always be updated together as they're part of the same repo with the same tags.
So to simplify the build process I made them one package.

**Comment by matteodelabre** [comment]
What’s the rationale behind switching to a binary package @timower? When possible, we prefer building from source.

File: package/timower-stuff/package
Comment (raisjn) [reply package/timower-stuff/package]: i think it's reasonable to keep as is (one package file with multiple packages out of the same repo), but that's because it's how i also do it

**Comment by timower** [comment]
> What’s the rationale behind switching to a binary package @timower? When possible, we prefer building from source.

Just so the build script is in one location. I'll update the pull request when I have time, for now I'll just leave it as a draft here.
(Also, I was looking into using GH actions to automatically create a PR when a tag gets pushed to my repo).

**Comment by Eeems** [comment]
@timower Any further progress on this? I'd love to have yaft added so I don't have to use fingerterm anymore :)

**Comment by timower** [comment]
> @timower Any further progress on this? I'd love to have yaft added so I don't have to use fingerterm anymore :)

I can update the PR this weekend probably. The only major issue right now is rm1 support (https://github.com/timower/rM2-stuff/issues/3)

**Comment by Eeems** [comment]
Hmm, well if I can get some time I might try to help with rM1 support. Since that's all I have.

**Comment by raisjn** [comment]
rM1 is now supported in yaft, so we should revisit this

File: package/yaft/package
Comment (timower) [inline package/yaft/package]: @matteodelabre It seems $pkgdir is not available during build, but I think it should be

**Review by raisjn — APPROVED** [review]
i tested this on rM1 and rM2 by installing it, then opening yaft and typing in a few commands / tab completions (cd, ls, etc) - works well, is nice and snappy, thank you for contributing it!

**Comment by Eeems** [comment]
There is no application registration file(s) included, so I can't launch it from oxide or remux. Are you planning on including one @timower?

**Comment by raisjn** [comment]
there are draft files and an icon, afaict, but not in the toltec PR. (and i think i saw a commit for them in timower's repo earlier today)

```
reMarkable: ~/ opkg files yaft
Package yaft (0.0.4-1) is installed on root and has the following files:
/opt/etc/draft/icons/yaft.png
/opt/share/terminfo/y/yaft-256color
/opt/etc/draft/yaft.draft
/opt/bin/yaft
reMarkable: ~/ cat /opt/etc/draft/yaft.draft
# Copyright (c) 2020 The Toltec Contributors
# SPDX-License-Identifier: MIT

name=yaft
desc=Framebuffer terminal emulator with an on-screen touch keyboard
call=/opt/bin/yaft
term=:
imgFile=yaft

```

**Comment by timower** [comment]
Indeed, I moved the draft files to my repo. So maintaining this package should be as easy as updating the sources url and checksum

**Pull request merged**
